matchers/matchers.go

```diff
@@ -0,0 +1,156 @@
+package matchers
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/a8m/expect"
+)
+
+// ViaPolling matches by polling the child matcher until it returns a sucess.
+// It can take polling options test length and rate. Test length is the first
+// argument and it determines how long to poll. Rate is the second argument
+// and it sets what is the frequency to poll. ViaPolling will return success
+// the first time the child matcher returns a success. If the child matcher
+// never returns a nil, then it will return the last error.
+func ViaPolling(m expect.Matcher, pollingOpts ...string) expect.Matcher {
+	return pollingMatcher{
+		childMatcher: m,
+		pollingOpts:  pollingOpts,
+	}
+}
+
+// Always matches by polling the child matcher until it returns an error.
+// It can take polling options test length and rate. Test length is the first
+// argument and it determines how long to poll. Rate is the second argument
+// and it sets what is the frequency to poll. Always will return an error
+// the first time the child matcher returns an error. If the child matcher
+// never returns an error, then it will return a nil.
+func Always(m expect.Matcher, pollingOpts ...string) expect.Matcher {
+	return pollingMatcher{
+		always:       true,
+		childMatcher: m,
+		pollingOpts:  pollingOpts,
+	}
+}
+
+// Match takes a value that can change over time. Therefore, the only
+// two valid options are a function with no arguments and a single return
+// type, or a readable channel. Anything else will return an error.
+//
+// If actual is a channel, then the child matcher will have to handle
+// reading from the channel.
+//
+// If the actual is a function, then the matcher will invoke the value
+// and pass the returned value to the child matcher.
+func (m pollingMatcher) Match(actual interface{}) error {
+	duration := time.Second
+	if m.always {
+		duration = 100 * time.Millisecond
+	}
+
+	rate := 10 * time.Millisecond
+
+	if len(m.pollingOpts) > 0 {
+		var err error
+		duration, err = time.ParseDuration(m.pollingOpts[0])
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(m.pollingOpts) > 1 {
+		var err error
+		rate, err = time.ParseDuration(m.pollingOpts[1])
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(m.pollingOpts) > 2 {
+		return fmt.Errorf("invalid number of args: %d", len(m.pollingOpts))
+	}
+
+	f, err := m.fetchFunc(actual)
+	if err != nil {
+		return err
+	}
+
+	if m.always {
+		return m.matchAlways(duration, rate, f)
+	}
+
+	return m.matchViaPolling(duration, rate, f)
+}
+
+type pollingMatcher struct {
+	always       bool
+	childMatcher expect.Matcher
+	pollingOpts  []string
+}
```

Comment:
I'd like to see separate `AlwaysMatcher` and `ViaPollingMatcher` structs (names pending), preferably exported (with exported fields) for anyone who doesn't want to use constructors.

The `pollingOpts` field should be split out into the actual options available (i.e. `Interval` and `Duration`).